### PR TITLE
Enable nullability in some Designer classes

### PR DESCRIPTION
**Pull request by gpetrou**
## Proposed changes

- Enable nullability in some Designer classes.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/10702)

### CODE REVIEW

**Comment by gpetrou** [comment]
@lonitra any thoughts on this one?

**Comment by SimonZhao888** [comment]
Looks good!

**Review by SimonZhao888 — DISMISSED** [review]

**Comment by lonitra** [comment]
Previously team had decided to avoid nullability changes in `System.Windows.Forms.Design` in favor of prioritizing nullability in `System.Windows.Forms` (https://github.com/dotnet/winforms/issues/8342#issuecomment-1769354111). Given that we are very close to completing nullability in `System.Windows.Forms`, I will revisit enabling nullability in `System.Windows.Forms.Design` with team.

**Comment by gpetrou** [comment]
> Previously team had decided to avoid nullability changes in `System.Windows.Forms.Design` in favor of prioritizing nullability in `System.Windows.Forms` ([#8342 (comment)](https://github.com/dotnet/winforms/issues/8342#issuecomment-1769354111)). Given that we are very close to completing nullability in `System.Windows.Forms`, I will revisit enabling nullability in `System.Windows.Forms.Design` with team.

There are only two files remaining to complete in `System.Windows.Forms` and there are open PRs about them. So, I thought about opening PRs for `System.Windows.Forms.Design`. I will not create new PRs, but perhaps we could merge the already open ones.

**Comment by lonitra** [comment]
> There are only two files remaining to complete in `System.Windows.Forms` and there are open PRs about them. So, I thought about opening PRs for `System.Windows.Forms.Design`. I will not create new PRs, but perhaps we could merge the already open ones.

Discussed with team and we will start taking a look at nullability changes in `System.Windows.Forms.Design` space again (https://github.com/dotnet/winforms/issues/8342#issuecomment-1917783656). However, generally our test coverage for this space is lacking, so in order to ensure that unintentional behavior changes are not made, we'd like basic test coverage for public methods to be added if the class is lacking coverage. Could you add some test coverage for the classes you are enabling nullability for in a separate PR?

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ButtonBaseDesigner.cs [hunk 1]
```diff
@@ -53,17 +51,17 @@ public override IList SnapLines
...
 
-            PropertyDescriptor prop;
+            PropertyDescriptor? prop;
             PropertyDescriptorCollection props = TypeDescriptor.GetProperties(Component);
 
             if ((prop = props["TextAlign"]) is not null)
             {
-                alignment = (ContentAlignment)prop.GetValue(Component);
+                alignment = (ContentAlignment)prop.GetValue(Component)!;
             }
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ButtonBaseDesigner.cs:60]: We seem to be doing this a lot across all the designer classes. Could we can create a helper method for this that also avoid using the `!` everywhere?

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/LabelDesigner.cs [hunk 2]
```diff
@@ -63,16 +61,16 @@ public override IList SnapLines
...
-                Type type = Type.GetType("System.Windows.Forms.Label");
+                Type? type = Type.GetType("System.Windows.Forms.Label");
                 if (type is not null)
                 {
-                    MethodInfo info = type.GetMethod("GetLeadingTextPaddingFromTextFormatFlags", BindingFlags.Instance | BindingFlags.NonPublic);
+                    MethodInfo? info = type.GetMethod("GetLeadingTextPaddingFromTextFormatFlags", BindingFlags.Instance | BindingFlags.NonPublic);
                     if (info is not null)
                     {
-                        int offset = (int)info.Invoke(Component, null);
+                        int offset = (int)info.Invoke(Component, parameters: null)!;
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/LabelDesigner.cs:73]: Please add a comment for this

**Review by lonitra — COMMENTED** [review]
Changes look good, it would be great if we could avoid all the `!` by creating a helper method

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ControlDesigner.cs [hunk 3]
```diff
@@ -2518,4 +2518,21 @@ internal void SetUnhandledException(Control owner, Exception exception)
 
         Control.Invalidate(true);
     }
+
+    internal static bool TryGetPropertyDescriptorValue<T>(
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ControlDesigner.cs:2522]: I hadn't realized that we already had existing methods that are similar. Could we make this into an extension method off of `PropertyDescriptor` and move it to https://github.com/dotnet/winforms/blob/00b11a73177daf5687673640b3629e51e7a9c4d3/src/System.Windows.Forms.Primitives/src/System/ComponentModel/MemberDescriptorExtensions.cs to keep things in the same place? I think we will still want an overload with a `ref` as you have here, but consider leveraging the other extension methods that are there

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ComboBoxDesigner.cs [hunk 4]
```diff
@@ -108,10 +108,10 @@ public override SelectionRules SelectionRules
             SelectionRules rules = base.SelectionRules;
             object component = Component;
 
-            PropertyDescriptor propStyle = TypeDescriptor.GetProperties(component)["DropDownStyle"];
+            PropertyDescriptor? propStyle = TypeDescriptor.GetProperties(component)["DropDownStyle"];
             if (propStyle is not null)
             {
-                ComboBoxStyle style = (ComboBoxStyle)propStyle.GetValue(component);
+                ComboBoxStyle style = (ComboBoxStyle)propStyle.GetValue(component)!;
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ComboBoxDesigner.cs:114]: Let's use one of the extension methods here

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/LabelDesigner.cs [hunk 5]
```diff
@@ -146,10 +148,10 @@ public override SelectionRules SelectionRules
             SelectionRules rules = base.SelectionRules;
             object component = Component;
 
-            PropertyDescriptor propAutoSize = TypeDescriptor.GetProperties(component)["AutoSize"];
+            PropertyDescriptor? propAutoSize = TypeDescriptor.GetProperties(component)["AutoSize"];
             if (propAutoSize is not null)
             {
-                bool autoSize = (bool)propAutoSize.GetValue(component);
+                bool autoSize = (bool)propAutoSize.GetValue(component)!;
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/LabelDesigner.cs:154]: Ditto here. There are other places in this change as well.

**Review by lonitra — APPROVED** [review]
LGTM, thank you!

**Pull request merged**